refactor: allow database init to recover from errors

crepererum opened the pull request:
Closes #3335.

alamb commented on server/src/database.rs:
```diff
@@ -715,7 +729,7 @@ impl Drop for Database {
 struct DatabaseShared {
     /// Configuration provided to the database at startup
-    config: DatabaseConfig,
+    config: RwLock<DatabaseConfig>,
```
I don't fully understand the need for a RwLock -- it seemed to me that one of the nice properties of the various database states was using the type system to encode the state transitions / when things could change

Allowing `config` to be changed seems to break that nice property

If you need to allow multiple readers, perhaps you could just use `config: Arc<DatabaseConfig>` rather than a lock?

crepererum replied:
The lock is here to allow the change of `skip_replay`. The old solution directly changed the state which for the new state machine is kinda tricky, because you need to hit the precise moment when the DB is in the `ReplayError` state. So instead we're now using the boolean flag to control this and allow the normal startup system to just use the flag to converge to the desired end state (in this case by skipping replay).

tustvold replied:
> because you need to hit the precise moment when the DB is in the ReplayError state

I don't follow why this is a problem, if you acquire a freeze handle you effectively prevent any concurrent transitions?

alamb commented on server/src/database.rs:
```diff
@@ -869,20 +879,29 @@ async fn background_worker(shared: Arc<DatabaseShared>) {
+    let db_name = shared.config.read().name.clone();
     info!(%db_name, "database initialization started");
 
+    // error throttle controshared.config.name.l
```
I don't understand this comment (what is a `controshared.config.name.l`?)

crepererum replied:
Probably a "paste random code here" error :sweat_smile:

crepererum replied:
fixed

alamb commented on server/src/database.rs:
```diff
@@ -891,50 +910,61 @@ async fn initialize_database(shared: &DatabaseShared) {
+
+                    // exponential backup w/ jitter, decorrelated
+                    // see https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+                    let mut rng = thread_rng();
+                    sleep = Duration::from_secs_f64(MAX_BACKOFF.as_secs_f64().min(
+                        rng.gen_range(INIT_BACKOFF.as_secs_f64()..(sleep.as_secs_f64() * 3.0)),
+                    ));
```
Given this has some magic numbers (and I am a big confused about the calculation for the backoff), what would you think about using a constructor on `TransactionOrWait` and encapsulating `TransationAndSleep` , and throttled. 

The logic might be easier to discover.

Maybe something like 
```
struct Backoff {
  sleep: Duration
}

impl Backoff {
  fn new() -> Self {
    // default backoff
   }
   fn backoff(self) -> Self {
      // put random backoff here
    }
}
```

Then instead of 
```rust
    let mut throttled_error = false;
    let mut sleep = INIT_BACKOFF;
``` 

In your main loop you could have something like

```rust
    let mut backoff: Option<Backoff> = None;
```

And update it in the loop

crepererum replied:
I've tried to improve the docs a bit and explain why not all wait operations are error backoffs.

alamb commented on server/src/database.rs:
```diff
@@ -891,50 +910,61 @@ async fn initialize_database(shared: &DatabaseShared) {
-                    None
+                    throttled_error = true;
+
+                    // exponential backup w/ jitter, decorrelated
```
```suggestion
                    // exponential backoff w/ jitter, decorrelated
```

alamb commented on server/src/database.rs:
```diff
@@ -891,50 +910,61 @@ async fn initialize_database(shared: &DatabaseShared) {
+                    // see https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+                    let mut rng = thread_rng();
+                    sleep = Duration::from_secs_f64(MAX_BACKOFF.as_secs_f64().min(
+                        rng.gen_range(INIT_BACKOFF.as_secs_f64()..(sleep.as_secs_f64() * 3.0)),
```
By using `INIT_BACKOFF .. sleep*3` doesn't that mean the backoff can actually decrease (rather than increase) each iteration?

I would have expected it to be something like `sleep = min(MAX_BACKOFF, sleep.exp(1 + rand()))` 

My brain may be all mushy today from reviewing so much disparate code though

crepererum replied:
I think you're observation is correct. However this should only happen w/ a low likelyhood and is done to decorrelate the retry events. (see linked AWS post)

alamb commented on server/src/database.rs:
```diff
@@ -943,40 +973,61 @@ async fn initialize_database(shared: &DatabaseShared) {
             }
-            state => unreachable!("{:?}", state),
+            DatabaseState::ReplayError(state, _) => {
+                warn!(%db_name, "throwing away loaded catalog to recover from replay error");
```
👍

alamb replied:
Maybe could put this `warn!` into the actual `rollback()` function to ensure it was called.

crepererum replied:
done

alamb commented on server/src/database.rs:
```diff
@@ -1060,19 +1111,45 @@ pub enum InitError {
+///                 [NoActiveDatabase]
+///                        ^
+///                        |
+/// (start)------------>[Known]<------------>[DatabaseObjectStoreLookupError]
```
I think the state machine actually looks more like this (the error state transitions go directly to the next stage rather than back

```
...
[Known]------------>[DatabaseObjectStoreLookupError]
   |                              |
   |                              |
   v                              |
[DatabaseObjectStoreFound] <------+
...
```

crepererum replied:
done

alamb commented on server/src/database.rs:
```diff
@@ -1081,25 +1158,19 @@ enum DatabaseState {
     RulesLoaded(DatabaseStateRulesLoaded),
     CatalogLoaded(DatabaseStateCatalogLoaded),
+
+    // Terminal state (positive)
```
```suggestion
    // Terminal state (success)
```

alamb commented on server/src/database.rs:
```diff
@@ -1081,25 +1158,19 @@ enum DatabaseState {
-    /// Terminal State
-    DatabaseObjectStoreLookupError(DatabaseStateKnown, Arc<InitError>),
-    /// Terminal State
+    // Terminal state (negative)
```
```suggestion
    // Terminal state (failure)
```

alamb reviewed (commented):
This PR is quite cool @crepererum  -- it is neat to see that the databases can get out of  error states. 👍 

I think this PR should really have some sort of testing as well

Perhaps you could write an end to end test 
1. creates a database
2. releases it
3. move away owner.pb
4. restarts it
5. waits for the db to get into `OwnerInfoLoadError`
6. puts owner.pb back
7. waits for the db to initialize successfully

I also had some questions about the use of a `RwLock` and the calculation of exponential backoff that I would like answered before I feel comfortable approving this PR.

crepererum commented:
@alamb I've added a test. I've also changed the state machine so it can also recover from `NoActiveDatabase` (there was no reason why it shouldn't be able to do so) and extended the `claim_database` test accordingly.

alamb commented on server/src/database.rs:
```diff
@@ -2101,6 +2101,41 @@ mod tests {
+        iox_object_store.put_owner_file(owner_backup).await.unwrap();
+        tokio::time::timeout(Duration::from_secs(10), async move {
+            loop {
+                if database.wait_for_init().await.is_ok() {
```
👍

alamb reviewed (approved):
Thank you @crepererum 


> The lock is here to allow the change of skip_replay. The old solution directly changed the state which for the new state machine is kinda tricky, because you need to hit the precise moment when the DB is in the ReplayError state. So instead we're now using the boolean flag to control this and allow the normal startup system to just use the flag to converge to the desired end state (in this case by skipping replay).

I still don't really like this but I admit I am now just whining as I don't have a better alternate suggestion.

alamb commented:
cc @tustvold you may be interested in this

tustvold commented on server/src/database.rs:
```diff
@@ -891,50 +920,53 @@ async fn initialize_database(shared: &DatabaseShared) {
                     match state.try_freeze() {
-                        Some(handle) => Some((DatabaseState::clone(&state), handle)),
+                        Some(handle) => {
+                            TransactionOrWait::Transaction(DatabaseState::clone(&state), handle)
```
It occurs to me that this dates from before you could wait on a freeze handle, we could probably remove this logic now

crepererum replied:
Done, it doesn't fully eliminate the variables though.

tustvold commented on server/src/database.rs:
```diff
@@ -869,20 +879,39 @@ async fn background_worker(shared: Arc<DatabaseShared>) {
 
+    // error throttling
+    // - checks if the current error was already throttled
+    // - keeps a backoff duration that will change over the curse of multiple errors
```
```suggestion
    // - keeps a backoff duration that will change over the course of multiple errors
```

tustvold commented on server/src/database.rs:
````diff
@@ -1060,19 +1108,55 @@ pub enum InitError {
-/// 1. It reaches `Initialized`
-///
-/// 2. It is reset to `Known` and starts initialization again
+/// ```text
````
:ok_hand:

tustvold reviewed (commented):
I think it is possible to remove the need for the new RWLock, along with the throttled_error complexity, but otherwise looks good

crepererum commented:
@tustvold simplified the locking and removed the user-visible state rollback (as we've agreed on during a 1:1 chat).

tustvold commented on server/src/database.rs:
```diff
@@ -676,30 +680,30 @@ impl Database {
 
     /// Recover from a ReplayError by skipping replay
     pub async fn skip_replay(&self) -> Result<(), Error> {
-        let db_name = &self.shared.config.name;
```
I don't think this method needs to be changed anymore - and therefore the RWLock can be removed?

crepererum replied:
You need it to trigger "skip replay" via gRPC. It's the only way to toggle the flag for a single database when an admin see it's broken and accepts that replay can be skipped for that particular database.

alamb commented:
THanks @crepererum
